bioneuron_oracle/builder.py

```diff
@@ -29,6 +31,22 @@ def __init__(self):
         self.spikes = neuron.h.Vector()
         self.ap_counter.record(neuron.h.ref(self.spikes))
         self.num_spikes_last = 0
+        self._clean = False
+
+    def update(self):
+        if self._clean:
+            raise RuntimeError("cannot update() after cleanup()")
+        count = len(self.spikes) - self.num_spikes_last
+        self.num_spikes_last = len(self.spikes)
+        return count, np.asarray(self.v_record)[-1]
+
+    def cleanup(self):
+        if self._clean:
+            raise RuntimeError("cleanup() may only be called once")
+        self.v_record.play_remove()
+        self.t_record.play_remove()
+        self.spikes.play_remove()
```

Comment:
`__exit__` is called when exiting from the `with` block, regardless of whether an error occurred. But I still need to check that the `play_remove` is doing what we want.

You could raise a `KeyboardInterrupt` in the middle of a `sim.run()` by calling it from within a node, but that makes no difference here or before.